source/extensions/filters/http/file_system_buffer/filter.h

```diff
@@ -1,26 +1,131 @@
+
 private:
+  // Merges any per-route config with the default config. Returns false if the config lacked
+  // a file manager.
+  bool initPerRouteConfig();
+  Http::FilterHeadersStatus onBadConfig();
+  Http::FilterDataStatus receiveData(BufferedStreamState& state, Buffer::Instance& data,
+                                     bool end_stream);
+  Http::FilterTrailersStatus receiveTrailers(const BufferBehavior& behavior,
+                                             BufferedStreamState& state);
+
+  // onStateChange makes stateful decisions about what to do next.
+  // Depending on the state of buffers and streams, may do one or more of the 'maybe' functions
+  // below.
+  void onStateChange();
+
+  // Conditional operations to perform on state change.
+  // If the server is ready to receive, sends any request data from the front of the buffer.
+  void maybeOutputRequest();
+  // If the client is ready to receive, sends any response data from the front of the buffer.
+  // If this finishes the connection, returns true.
+  bool maybeOutputResponse();
+  // If the buffer has too much queued for writing to disk, asks the source to slow down.
+  // If a low watermark was sent (that might require us to do another action), returns true.
+  bool maybeSendWatermarkUpdates();
+  // If the memory buffer is overfull, moves the *latest* chunks from memory to storage.
+  // If the memory buffer has room, moves the *earliest* chunks from storage to memory.
+  // These operations are asynchronous; the impacted buffer fragments are in an unusable state until
+  // the operation completes.
+  bool maybeStorage(BufferedStreamState& state, Http::StreamFilterCallbacks& callbacks);
+  std::function<void(absl::Status)> getOnFileActionCompleted();
+
+  // Called if an unrecoverable error occurs in the filter (e.g. a file operation fails). Internal
+  // server error.
+  void filterError(absl::string_view err);
+
+  // Returns a safe dispatch function that aborts if the filter has been destroyed.
+  std::function<void(std::function<void()>)> getSafeDispatch();
+
+  // Queue an onStateChange in the dispatcher. This is used to get the next piece of work back
+  // into the Envoy thread from an AsyncFiles thread, or to queue work that may not be allowed
+  // in the current Envoy context (e.g. updating watermark callbacks during a watermark callback
+  // is not allowed, injecting data to a stream during decodeData is not allowed).
+  void dispatchStateChanged();
+
+  BufferedStreamState request_state_;
+  BufferedStreamState response_state_;
+  Http::RequestHeaderMap* request_headers_;
+  Http::ResponseHeaderMap* response_headers_;
+
   const std::shared_ptr<FileSystemBufferFilterConfig> base_config_;
+  // config_ is a merge of the base config and any per-route config.
+  absl::optional<FileSystemBufferFilterMergedConfig> config_;
+  Http::StreamDecoderFilterCallbacks* request_callbacks_;
+  Http::StreamEncoderFilterCallbacks* response_callbacks_;
```

Comment:
nit: I would initialize to nullptr since they don't get initialized until after the constructor runs.